addons/membership/models/partner.py

```diff
@@ -0,0 +1,173 @@
+                FROM
+                    res_partner p
+                LEFT JOIN
+                    membership_membership_line m
+                    ON (m.partner_id = p.id)
+                WHERE
+                    p.id IN %s
+                GROUP BY
+                    p.id""", (tuple(ids_to_search), ))
+            for record in self.env.cr.dictfetchall():
+                partner = self.browse(record.pop('id')).update(record)
+
+    @api.depends('member_lines_ids.account_invoice_id.state', 'membership_state', 'associate_member_id', 'free_member')
+    def _compute_membership_state(self):
+        return self._membership_state()
+
+    @api.constrains('associate_member_id')
+    def _check_recursion(self):
+        """Check  Recursive  for Associated Members.
+        """
+        level = 100
+        recursive_ids = self.ids
+        while len(recursive_ids):
+            self.env.cr.execute('SELECT DISTINCT associate_member_id FROM res_partner WHERE id IN %s', (tuple(recursive_ids),))
+            recursive_ids = filter(None, map(lambda x: x[0], self.env.cr.fetchall()))
+            if not level:
+                raise UserError(_('Error ! You cannot create recursive associated members.'))
+            level -= 1
+        return
+
+    @api.model
+    def _cron_update_membership(self):
+        partners = self.search([('membership_state', '=', 'paid')])
+        if partners:
+            partners._store_set_values(['membership_state'])
+
+    @api.multi
+    def _membership_state(self):
+        """This Function return Membership State For Given Partner. """
+        today_date = date.today()
+        for partner_data in self:
+            partner_data.membership_state = 'none'
+            if partner_data.membership_cancel and today_date > fields.Date.from_string(partner_data.membership_cancel):
+                partner_data.membership_state = 'free' if partner_data.free_member else 'canceled'
+                continue
+            if partner_data.membership_stop and today_date > fields.Date.from_string(partner_data.membership_stop):
+                partner_data.membership_state = 'free' if partner_data.free_member else 'old'
+                continue
+
+            state = 4
+            if partner_data.member_lines_ids:
+                for mline in partner_data.member_lines_ids:
+                    if fields.Date.from_string(mline.date_to) >= today_date:
+                        if mline.account_invoice_line_id and mline.account_invoice_line_id.invoice_id:
+                            mstate = mline.account_invoice_line_id.invoice_id.state
+                            if mstate == 'paid':
+                                state = 0
+                                inv = mline.account_invoice_line_id.invoice_id
+                                for payment in inv.payment_ids:
+                                    if payment.invoice_ids and any(inv.type == 'out_refund' for inv in payment.invoice_ids):
```

Comment:
instead of above three lines, you can write 
```python
if mline.account_invoice_line_id.invoice_id.mapped('payment_ids.invoice_ids').filtered(lambda inv: inv.type == 'out_refund'):
```
And, check whether `mline.account_invoice_line_id.invoice_id` is same as `mline.account_invoice_id` or not. If they are same , you can use the latter one for this condition(only if you are pretty sure about it).